### PR TITLE
Add TensorBoard integration.

### LINKED ISSUE #1060: [Proposal] Add integration with Tensorboard
## Motivation
Tensorboard provides [HParams dashboard](https://www.tensorflow.org/tensorboard/hyperparameter_tuning_with_hparams) which is useful for identifying most promising sets of hyperparameters. It would be nice add custom module to optuna for writing experiment results to tensorboard.

## Description
We can add a callback, which writes experiment results to tensorboard. Below is an example of tuning number of units in a hidden layer, dropout rate and optimizer for a small network on fashion MNIST.

```
import optuna

import tensorflow as tf
from tensorboard.plugins.hparams import api as hp

fashion_mnist = tf.keras.datasets.fashion_mnist

(x_train, y_train),(x_test, y_test) = fashion_mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

HP_NUM_UNITS = hp.HParam('num_units', hp.Discrete([16, 32]))
HP_DROPOUT = hp.HParam('dropout', hp.RealInterval(0.1, 0.2))
HP_OPTIMIZER = hp.HParam('optimizer', hp.Discrete(['adam', 'sgd']))

METRIC_ACCURACY = 'accuracy'

with tf.summary.create_file_writer('logs/hparam_tuning').as_default():
    hp.hparams_config(
        hparams=[HP_NUM_UNITS, HP_DROPOUT, HP_OPTIMIZER],
        metrics=[hp.Metric(METRIC_ACCURACY, display_name='Accuracy')],
    )

def train_test_model(hparams):
    model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(hparams[HP_NUM_UNITS], activation=tf.nn.relu),
        tf.keras.layers.Dropout(hparams[HP_DROPOUT]),
        tf.keras.layers.Dense(10, activation=tf.nn.softmax),
    ])
    model.compile(
        optimizer=hparams[HP_OPTIMIZER],
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'],
    )

    model.fit(x_train, y_train, epochs=1) # Run with 1 epoch to speed things up for demo purposes
    _, accuracy = model.evaluate(x_test, y_test)
    return accuracy

def run(run_dir, hparams):
    with tf.summary.create_file_writer(run_dir).as_default():
        hp.hparams(hparams)  # record the values used in this trial
        accuracy = train_test_model(hparams)
        tf.summary.scalar(METRIC_ACCURACY, accuracy, step=1)

session_num = 0

def tensorboard_callback(study, trial):
    trial_value = trial.value if trial.value is not None else float("nan")

    num_units = trial.params['NUM_UNITS']
    dropout_rate = trial.params['DROPOUT_RATE']
    optimizer = trial.params['OPTIMIZER']

    hparams = {
        HP_NUM_UNITS: num_units,
        HP_DROPOUT: dropout_rate,
        HP_OPTIMIZER: optimizer,
    }

    global session_num
    run_name = "run-%d" % session_num
    run_dir = 'logs_optuna/hparam_tuning/' + run_name
    #run_dir = 'logs_optuna/hparam_tuning/'

    with tf.summary.create_file_writer(run_dir).as_default():
        hp.hparams(hparams)  # record the values used in this trial
        tf.summary.scalar(METRIC_ACCURACY, trial_value, step=1)
    session_num += 1

def objective(trial):

    num_units = trial.suggest_int('NUM_UNITS', HP_NUM_UNITS.domain.values[0], HP_NUM_UNITS.domain.values[1])
    dropout_rate = trial.suggest_uniform('DROPOUT_RATE', HP_DROPOUT.domain.min_value, HP_DROPOUT.domain.max_value)
    optimizer = trial.suggest_categorical('OPTIMIZER', HP_OPTIMIZER.domain.values)

    hparams = {
        HP_NUM_UNITS: num_units,
        HP_DROPOUT: dropout_rate,
        HP_OPTIMIZER: optimizer,
    }

    accuracy = train_test_model(hparams)
    return accuracy

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=10, timeout=600, callbacks=[tensorboard_callback])

```
Here is the result in Tensorboard HParams dashboard
image

### PR DESCRIPTION
**Pull request by VladSkripniuk**
fix #1060 .
## Description of the changes

This PR introduces `optuna.integration.TensorBoardCallback`. This callback saves values of objective function and parameters to [HParams](https://www.tensorflow.org/tensorboard/hyperparameter_tuning_with_hparams) dashboard.

### CODE REVIEW

**Comment by VladSkripniuk** [comment]
To let Tensorboard know the range of values for parameters I use `param_distributions` parameter in `TensorBoardCallback.__init__`. I think we need to discuss if we can somehow pass the distribution from `trial.suggest` to the callback.

File: optuna/integration/tensorboard.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+class TensorBoardCallback(object):
+    """Callback to track Optuna trials with TensorBoard.
+
+    This callback adds relevant information that is tracked by Optuna to TensorBoard.
+    Args:
+        metric_name:
+            Name of the metric. Since the metric itself is just a number,
+            `metric_name` can be used to give it a name. So you know later
+            if it was roc-auc or accuracy.
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:43]: Please specify whole information of all arguments. (`dirname ` and `param_distributions`)

File: optuna/integration/tensorboard.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    """
+
+    def __init__(self, dirname, param_distributions, metric_name):
+        # type: (str, Dict[str, optuna.distributions.BaseDistribution], str) -> None
+
+        _check_tensorboard_availability()
+
+        self._dirname = dirname
+        self._metric_name = metric_name
+        self._hp_params = dict()  # type: Dict
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:53]: We can specify its type in more detail. 
```suggestion
        self._hp_params = dict()  # type: Dict[str, hp.HParam]
```

File: examples/tensorboard_simple.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+fashion_mnist = tf.keras.datasets.fashion_mnist
+
+(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
+x_train, x_test = x_train / 255.0, x_test / 255.0
+
+
+def train_test_model(num_units, dropout_rate, optimizer):
+    # type: (int, float, str) -> float
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:18]: We can use the new style for the type hinting. For example, please refer to [here](https://github.com/optuna/optuna/blob/master/optuna/pruners/hyperband.py).

File: examples/tensorboard_simple.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import optuna
+from optuna.integration.tensorboard import TensorBoardCallback
+from optuna import type_checking
+import tensorflow as tf
+
+if type_checking.TYPE_CHECKING:
+    from typing import Dict  # NOQA
+    from typing import Any  # NOQA
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:8]: In the new type hinting style, we can remove the `type_checking`, just import `typing/Dict or Any` outside `if type_checking.TYPE_CHECKING:` and remove NOQAs.

File: optuna/integration/tensorboard.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
+import optuna
+from optuna import type_checking
+import os
+
+if type_checking.TYPE_CHECKING:
+    from typing import Any  # NOQA
+    from typing import Dict  # NOQA
+    from typing import Optional  # NOQA
+    from typing import Tuple  # NOQA
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:9]: Ditto

**Comment by HideakiImamura** [comment]
Hi @VladSkripniuk. I would like to discuss the user interface of `TensorBoardCallback`. I think that we can remove the `param_distributions` argument by automatically looking up the first finished trial. And it might be possible to add parameters by looking up finished trials automatically. What do you think?

File: optuna/integration/tensorboard.py (hunk 5 above)
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:9]: Please import build-in libraries at first, then import third-party libraries.

```suggestion
import os
from typing import Dict

import optuna
```

File: optuna/integration/tensorboard.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+def _check_tensorboard_availability() -> None:
+
+    if not _available:
+        raise ImportError(
+            "TensorBoard is not available. Please install TensorBoard to use this "
+            "feature. It can be installed by executing `$ pip install "
+            "tensorboard`. For further information, please refer to the installation guide "
+            "of TensorBoard. (The actual import error is as follows: " + str(_import_error) + ")"
+        )
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:33]: We can use DRY import. This feature is recently introduced in PR #1315.  

```suggestion
from optuna._imports import try_import

with try_import() as _imports:
    from tensorboard.plugins.hparams import api as hp
    import tensorflow as tf
```

File: optuna/integration/tensorboard.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            Name of the metric. Since the metric itself is just a number,
+            `metric_name` can be used to give it a name. So you know later
+            if it was roc-auc or accuracy.
+    """
+
+    def __init__(
+        self, dirname: "str", param_distributions: "Dict[str, Tuple[str, Any]]", metric_name: "str"
+    ) -> None:
+
+        _check_tensorboard_availability()
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:59]: By using DRY import, please replace it as follows.

```suggestion
        _imports.check()
```

File: optuna/integration/tensorboard.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        hparams = dict()
+        for param_name, param_value in trial.params.items():
+            hparams[self._hp_params[param_name]] = param_value
+
+        run_name = "trial-%d" % trial.number
+        run_dir = os.path.join(self._dirname, run_name)
+
+        with tf.summary.create_file_writer(run_dir).as_default():
+            hp.hparams(hparams)  # record the values used in this trial
+            tf.summary.scalar(self._metric_name, trial_value, step=1)
```
Comment (HideakiImamura) [inline optuna/integration/tensorboard.py:155]: We can remove the `param_distributions` argument by introducing "lazy" initialization as follows.

```suggestion
    def __init__(self, dirname: str, metric_name: str) -> None:

        _imports.check()

        self._dirname = dirname
        self._metric_name = metric_name
        self._hp_params = dict()  # type: Dict[str, hp.HParam]

    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:

        if len(self._hp_params) == 0:
            self._initialization(study)

        trial_value = trial.value if trial.value is not None else float("nan")

        hparams = dict()
        for param_name, param_value in trial.params.items():
            if param_name not in self._hp_params:
                self._add_distributions(trial.distributions)
            hparams[self._hp_params[param_name]] = param_value

        run_name = "trial-%d" % trial.number
        run_dir = os.path.join(self._dirname, run_name)

        with tf.summary.create_file_writer(run_dir).as_default():
            hp.hparams(hparams)  # record the values used in this trial
            tf.summary.scalar(self._metric_name, trial_value, step=1)

    def _add_distributions(
        self, distributions: Dict[str, optuna.distributions.BaseDistribution]
    ) -> None:

        for param_name, param_distribution in distributions.items():
            if isinstance(param_distribution, optuna.distributions.UniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.LogUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.DiscreteUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.Discrete(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.IntUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.IntInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.CategoricalDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.Discrete(param_distribution.choices)
                )
            else:
                distribution_list = [
                    optuna.distributions.UniformDistribution.__name__,
                    optuna.distributions.LogUniformDistribution.__name__,
                    optuna.distributions.DiscreteUniformDistribution.__name__,
                    optuna.distributions.IntUniformDistribution.__name__,
                    optuna.distributions.CategoricalDistribution.__name__,
                ]
                raise NotImplementedError(
                    "The distribution {} is not implemented. "
                    "The parameter distribution should be one of the {}".format(
                        param_distribution, distribution_list
                    )
                )

    def _initialization(self, study: optuna.Study) -> None:

        completed_trials = [
            trial
            for trial in study.get_trials(deepcopy=False)
            if trial.state == optuna.trial.TrialState.COMPLETE
        ]
        for trial in completed_trials:
            self._add_distributions(trial.distributions)
```

File: examples/tensorboard_simple.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    _, accuracy = model.evaluate(x_test, y_test)
+    return accuracy
+
+
+param_distributions = dict()  # type: Dict[str, Any]
+param_distributions["NUM_UNITS"] = ("int_uniform", (16, 32))
+param_distributions["DROPOUT_RATE"] = ("uniform", (0.1, 0.2))
+param_distributions["OPTIMIZER"] = ("categorical", (["sgd", "adam"],))
+
+
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:40]: By introducing lazy initialization, we can omit them.

```suggestion
```

File: examples/tensorboard_simple.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    num_units = trial.suggest_int(
+        "NUM_UNITS", param_distributions["NUM_UNITS"][1][0], param_distributions["NUM_UNITS"][1][1]
+    )
+    dropout_rate = trial.suggest_uniform(
+        "DROPOUT_RATE",
+        param_distributions["DROPOUT_RATE"][1][0],
+        param_distributions["DROPOUT_RATE"][1][1],
+    )
+    optimizer = trial.suggest_categorical("OPTIMIZER", param_distributions["OPTIMIZER"][1][0])
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:51]: By introducing lazy initialization, we need to give arguments explicitly.

```suggestion
    num_units = trial.suggest_int("NUM_UNITS", 16, 32)
    dropout_rate = trial.suggest_uniform("DROPOUT_RATE", 0.1, 0.2)
    optimizer = trial.suggest_categorical("OPTIMIZER", ["sgd", "adam"])
```

File: examples/tensorboard_simple.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    accuracy = train_test_model(num_units, dropout_rate, optimizer)  # type: ignore
+    return accuracy
+
+
+tensorboard_callback = TensorBoardCallback("logs/", param_distributions, metric_name="accuracy")
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:57]: By using lazy initialization, we can omit the second argument.

```suggestion
tensorboard_callback = TensorBoardCallback("logs/", metric_name="accuracy")
```

File: tests/integration_tests/test_tensorboard.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def test_study_name() -> None:
+
+    distributions = {
+        "x": ("uniform", (-1.0, 1.0)),
+        "y": ("loguniform", (20.0, 30.0)),
+        "z": ("categorical", ((-1.0, 1.0),)),
+    }
+
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:28]: By using lazy distributions, we can remove them.

```suggestion
```

File: tests/integration_tests/test_tensorboard.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import optuna
+from optuna.integration.tensorboard import TensorBoardCallback
+
+from tensorboard.backend.event_processing.event_accumulator import EventAccumulator
+
+
+def _objective_func(trial: "optuna.trial.Trial") -> float:
+
+    x = trial.suggest_uniform("x", -1.0, 1.0)
+    y = trial.suggest_loguniform("y", 20, 30)
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:14]:
```suggestion
    y = trial.suggest_loguniform("y", 20.0, 30.0)
```

File: tests/integration_tests/test_tensorboard.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        "x": ("uniform", (-1.0, 1.0)),
+        "y": ("loguniform", (20.0, 30.0)),
+        "z": ("categorical", ((-1.0, 1.0),)),
+    }
+
+    dirname = tempfile.mkdtemp()
+    metric_name = "target"
+    study_name = "test_tensorboard_integration"
+
+    tbcallback = TensorBoardCallback(dirname, distributions, metric_name)
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:33]:
```suggestion
    tbcallback = TensorBoardCallback(dirname, metric_name)
```

**Comment by VladSkripniuk** [comment]
Hi @HideakiImamura thank for your great help! I've applied proposed changes.

File: tests/integration_tests/test_tensorboard.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import optuna
+from optuna.integration.tensorboard import TensorBoardCallback
+
+from optuna._imports import try_import
+
+with try_import() as _imports:
+    from tensorboard.backend.event_processing.event_accumulator import EventAccumulator
+
+
+def _objective_func(trial: "optuna.trial.Trial") -> float:
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:16]: Ditto.
```suggestion
def _objective_func(trial: optuna.trial.Trial) -> float:
```

File: examples/tensorboard_simple.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        optimizer=optimizer, loss="sparse_categorical_crossentropy", metrics=["accuracy"],
+    )
+
+    model.fit(x_train, y_train, epochs=1)  # Run with 1 epoch to speed things up for demo purposes
+    _, accuracy = model.evaluate(x_test, y_test)
+    return accuracy
+
+
+def objective(trial: optuna.trial.Trial) -> float:
+
```
Comment (HideakiImamura) [inline examples/tensorboard_simple.py:31]: We can just remove this blank line :).
```suggestion
```

File: tests/integration_tests/test_tensorboard.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import optuna
+from optuna._imports import try_import
+from optuna.integration.tensorboard import TensorBoardCallback
+
+with try_import() as _imports:
+    from tensorboard.backend.event_processing.event_accumulator import EventAccumulator
+
+
+def _objective_func(trial: optuna.trial.Trial) -> float:
+
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:16]: Ditto.
```suggestion
```

File: tests/integration_tests/test_tensorboard.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    assert isinstance(z, float)
+    trial.set_user_attr("my_user_attr", "my_user_attr_value")
+    return (x - 2) ** 2 + (y - 25) ** 2 + z
+
+
+@pytest.mark.skipif(
+    sys.version_info > (3, 7), reason="Tensorflow is not installed for python 3.8."
+)
+def test_study_name() -> None:
+
```
Comment (HideakiImamura) [inline tests/integration_tests/test_tensorboard.py:29]: Ditto.
```suggestion
```

**Pull request merged**